File: client-go/transaction_builder.go
package queen

import (
	"context"
	"fmt"
)

// TransactionBuilder provides a fluent API for atomic operations.
type TransactionBuilder struct {
	httpClient     *HttpClient
	operations     []Operation
	requiredLeases map[string]bool
}

// NewTransactionBuilder creates a new TransactionBuilder.
func NewTransactionBuilder(httpClient *HttpClient) *TransactionBuilder {
	return &TransactionBuilder{
		httpClient:     httpClient,
		operations:     make([]Operation, 0),
		requiredLeases: make(map[string]bool),
	}
}

// Ack adds an acknowledgment operation to the transaction.
// messages can be *Message, []*Message, or []Message
// status should be "completed" or "failed"
func (tb *TransactionBuilder) Ack(messages interface{}, status string, opts AckOptions) *TransactionBuilder {
	// Normalize messages to slice
	var msgs []*Message
	switch m := messages.(type) {
	case *Message:
		msgs = []*Message{m}
	case []*Message:
		msgs = m
	case []Message:
		for i := range m {
			msgs = append(msgs, &m[i])
		}
	case Message:
		msgs = []*Message{&m}
	}

	// Add ack operations
	for _, msg := range msgs {
		op := Operation{
			Type:          "ack",
			TransactionID: msg.TransactionID,
			PartitionID:   msg.PartitionID,
			Status:        status,
			ConsumerGroup: opts.ConsumerGroup,
		}
		tb.operations = append(tb.operations, op)

		// Track required leases
		if msg.LeaseID != "" {
			tb.requiredLeases[msg.LeaseID] = true
		}
	}

	return tb
}

// Queue returns a TransactionQueueBuilder for adding push operations.
func (tb *TransactionBuilder) Queue(name string) *TransactionQueueBuilder {
	return &TransactionQueueBuilder{
		tb:        tb,
		queueName: name,
		partition: DefaultPartition,
	}
}

// Commit executes the transaction atomically.
func (tb *TransactionBuilder) Commit(ctx context.Context) (*TransactionResponse, error) {
	if len(tb.operations) == 0 {
		return nil, fmt.Errorf("transaction has no operations")
	}

	// Build required leases list
	leases := make([]string, 0, len(tb.requiredLeases))
	for lease := range tb.requiredLeases {
		leases = append(leases, lease)
	}

	// Build request
	req := transactionRequest{
		Operations:     tb.operations,
		RequiredLeases: leases,
	}

	// Make request
	result, err := tb.httpClient.Post(ctx, "/api/v1/transaction", req)
	if err != nil {
		return nil, fmt.Errorf("transaction commit failed: %w", err)
	}

	// Parse response
	response := &TransactionResponse{Success: true}
	if success, ok := result["success"].(bool); ok {
		response.Success = success
	}
	if errMsg, ok := result["error"].(string); ok && errMsg != "" {
		response.Success = false
		response.Error = errMsg
	}

	logInfo("TransactionBuilder.Commit", map[string]interface{}{
		"operations": len(tb.operations),
		"leases":     len(leases),
		"success":    response.Success,
	})

	return response, nil
}

// TransactionQueueBuilder provides a fluent API for adding push operations to a transaction.
type TransactionQueueBuilder struct {
	tb        *TransactionBuilder
	queueName string
	partition string
}

// Partition sets the partition for push operations.
func (tqb *TransactionQueueBuilder) Partition(name string) *TransactionQueueBuilder {
	if name == "" {
		tqb.partition = DefaultPartition
	} else {
		tqb.partition = name
	}
	return tqb
}

// Push adds push operations to the transaction.
// payload can be a single item or a slice.
func (tqb *TransactionQueueBuilder) Push(payload interface{}) *TransactionBuilder {
	// Normalize payload to slice
	var payloads []interface{}
	switch p := payload.(type) {
	case []interface{}:
		payloads = p
	case []map[string]interface{}:
		for _, m := range p {
			payloads = append(payloads, m)
		}
	default:
		payloads = []interface{}{payload}
	}

	// Build push items
	items := make([]PushItem, len(payloads))
	for i, p := range payloads {
		items[i] = PushItem{
			Queue:         tqb.queueName,
			Partition:     tqb.partition,
			Payload:       p,
			TransactionID: GenerateUUID(),
		}
	}

	// Add push operation
	op := Operation{
		Type:  "push",
		Items: items,
	}
	tqb.tb.operations = append(tqb.tb.operations, op)

	return tqb.tb
}

File: client-laravel/src/Builders/TransactionBuilder.php
<?php

namespace Queen\Builders;

use Queen\Http\HttpClient;

class TransactionBuilder
{
    private HttpClient $httpClient;
    private array $operations = [];
    private array $requiredLeases = [];

    public function __construct(HttpClient $httpClient)
    {
        $this->httpClient = $httpClient;
    }

    public function ack(array|object $messages, string $status = 'completed', array $context = []): static
    {
        $msgs = is_array($messages) && (isset($messages[0]) || empty($messages)) ? $messages : [$messages];

        foreach ($msgs as $msg) {
            if (is_string($msg)) {
                $transactionId = $msg;
                $partitionId = null;
                $leaseId = null;
            } else {
                $msg = (array) $msg;
                $transactionId = $msg['transactionId'] ?? $msg['id'] ?? null;
                $partitionId = $msg['partitionId'] ?? null;
                $leaseId = $msg['leaseId'] ?? null;
            }

            if ($transactionId === null) {
                throw new \InvalidArgumentException('Message must have transactionId or id property');
            }

            if ($partitionId === null) {
                throw new \InvalidArgumentException('Message must have partitionId property to ensure message uniqueness');
            }

            $operation = [
                'type' => 'ack',
                'transactionId' => $transactionId,
                'partitionId' => $partitionId,
                'status' => $status,
            ];

            if (isset($context['consumerGroup'])) {
                $operation['consumerGroup'] = $context['consumerGroup'];
            }

            $this->operations[] = $operation;

            if ($leaseId !== null) {
                $this->requiredLeases[] = $leaseId;
            }
        }

        return $this;
    }

    /**
     * Returns a sub-builder for push operations on a queue
     */
    public function queue(string $queueName): TransactionQueueBuilder
    {
        return new TransactionQueueBuilder($this, $queueName);
    }

    /**
     * @internal Used by TransactionQueueBuilder
     */
    public function addPushOperation(string $queueName, ?string $partition, array $items): void
    {
        $this->operations[] = [
            'type' => 'push',
            'items' => array_map(function (array $item) use ($queueName, $partition) {
                $payloadValue = $item['data'] ?? $item['payload'] ?? $item;

                $result = [
                    'queue' => $queueName,
                    'payload' => $payloadValue,
                ];

                if ($partition !== null) {
                    $result['partition'] = $partition;
                }

                return $result;
            }, $items),
        ];
    }

    public function commit(): array
    {
        if (empty($this->operations)) {
            throw new \RuntimeException('Transaction has no operations to commit');
        }

        $result = $this->httpClient->post('/api/v1/transaction', [
            'operations' => $this->operations,
            'requiredLeases' => array_values(array_unique($this->requiredLeases)),
        ]);

        if (!($result['success'] ?? false)) {
            $error = $result['error'] ?? 'Transaction failed';
            $txId = $result['transactionId'] ?? null;
            $msg = $txId ? "Transaction {$txId} failed: {$error}" : "Transaction failed: {$error}";
            throw new \RuntimeException($msg);
        }

        return $result;
    }
}

class TransactionQueueBuilder
{
    private TransactionBuilder $parent;
    private string $queueName;
    private ?string $partition = null;

    public function __construct(TransactionBuilder $parent, string $queueName)
    {
        $this->parent = $parent;
        $this->queueName = $queueName;
    }

    public function partition(string $partition): static
    {
        $this->partition = $partition;
        return $this;
    }

    public function push(array $items): TransactionBuilder
    {
        $this->parent->addPushOperation($this->queueName, $this->partition, $items);
        return $this->parent;
    }
}
